### DISCUSSION TITLE
Add signals to Camera nodes triggered when they gain or lose "current" status

### DISCUSSION THREAD

**Discussion by mrkdji** [godotengine/godot-proposals#9089] [Engine Core]
I am working on a point & click game with multiple cameras and it would have been convenient to have something like **current_status_changed(value)** to activate / deactivate input processing.

**Comment by AThousandShips** [comment]
When does camera switching happen automatically? If it's controlled by code you can just signal when you change it, no need for a signal, signals are meant for things that happen outside of your control

I'd instead suggest a notification, internally `Camera3D` already has one, so adding one for the 2D case and also exposing them both would solve this without the complexity of a signal, and it'd be more appropriate as it's the camera that should respond to this change IMO, not some other node

**Reply by Mickeon** [reply]
It's likely that whatever the solution is, OP would be satisfied if there was _any_ way to detect this outside of the code handling `make_current`.

**Reply by mrkdji** [reply]
In my case camera switching depends on the player clicking on certain Area3D.
I use signals also to make my code more self-contained and have some nodes manage themselves.
I am aware this is totally doable with a few lines of code, like: 
- create a new class extending Camera3D that register to a custom signal in its _ready;
- create possibly an autoload for camera switching that emits a signal whenever a set_main_camera method is called;
- Do something if the newly set camera is == self inside the extended camera code

Having a signal for this would simply be more convenient.

To be fair exposing a notification would be ok, but if one were to expose the notification I cannot see the downsides of exposing a new signal, it'd pretty much work like Control nodes focus_entered focus_exited.

**Reply by AThousandShips** [reply]
If there's a notification you can handle it with a custom signal if you want, but a signal needs to be needed, signals clutter and if few use them they're just bloating the engine, we add things only if they are necessary

A notification is more contained, a local solution, and has less overhead

The need for this seems very limited generally, so adding a whole signal is a bit much IMO, a notification would be more justified for this case

But for your use case there still isn't a strict need for either, this isn't a situation where you *can't* work around this, so that's important to keep in mind

The only time, AFAIK, that a camera can become current *without* you knowing it will be so is when some other camera is current and is turned off or removed, so there's a reason to add some feature here, but I'd say it should be a notification as it's the most appropriate

**Reply by mrkdji** [reply]
A notification is perfectly fine for what I want to achieve, and it would be much better than nothing, but I still believe that a signal would be more beginner friendly.

**Reply by Mickeon** [reply]
> To be fair exposing a notification would be ok, but if one were to expose the notification I cannot see the downsides of exposing a new signal, it'd pretty much work like Control nodes focus_entered focus_exited.

Signals do not come for free. Even if the signal is not connected to anything there's always a base performance cost, albeit relatively small. The same goes for notifications, but those are often very much necessary internally.
However, I don't necessarily disagree. It's not like switching cameras happens ridiculously often. The important question is always "how often will developers make use of this". So, good call opening a discussion about this. I would recommend making a full proposal to catch more eyes, as well.

**Reply by AThousandShips** [reply]
*If* we add a signal I'd add it to `viewport` instead and have it convey the new current camera, that seems more useful for a signal, having a local signal on each camera means a notification works just as well and I'd say that would be the most appropriate change

**Reply by Mickeon** [reply]
Yeah that's sound to me. If there's no camera available at all the Viewport signal would just emit the signal and pass `null`.
Perhaps the signal should instead pass the _previously active_ camera, instead? Just food for thought. It's why it'd be nice to think of a good proposal.

**Reply by AThousandShips** [reply]
I'd say to pass both `camera_2d/3d_changed(old, new)`
